packages/vite/src/node/plugins/html.ts

```diff
@@ -143,83 +136,139 @@ export const isAsyncScriptMap = new WeakMap<
   Map<string, boolean>
 >()
 
+export function nodeIsElement(
+  node: DefaultTreeAdapterMap['node']
+): node is DefaultTreeAdapterMap['element'] {
+  return node.nodeName[0] !== '#'
+}
+
+function traverseNodes(
+  node: DefaultTreeAdapterMap['node'],
+  visitor: (node: DefaultTreeAdapterMap['node']) => void
+) {
+  visitor(node)
+  if (
+    nodeIsElement(node) ||
+    node.nodeName === '#document' ||
+    node.nodeName === '#document-fragment'
+  ) {
+    node.childNodes.forEach((childNode) => traverseNodes(childNode, visitor))
+  }
+}
+
 export async function traverseHtml(
   html: string,
   filePath: string,
-  visitor: NodeTransform
+  visitor: (node: DefaultTreeAdapterMap['node']) => void
 ): Promise<void> {
   // lazy load compiler
-  const { parse, transform } = await import('@vue/compiler-dom')
-  // @vue/compiler-core doesn't like lowercase doctypes
-  html = html.replace(/<!doctype\s/i, '<!DOCTYPE ')
-  try {
-    const ast = parse(html, { comments: true })
-    transform(ast, {
-      nodeTransforms: [visitor]
-    })
-  } catch (e) {
-    handleParseError(e, html, filePath)
-  }
+  const { parse } = await import('parse5')
+  const ast = parse(html, {
+    sourceCodeLocationInfo: true,
+    onParseError: (e: ParserError) => {
+      handleParseError(e, html, filePath)
+    }
+  })
+  traverseNodes(ast, visitor)
 }
 
-export function getScriptInfo(node: ElementNode): {
-  src: AttributeNode | undefined
+export function getScriptInfo(node: DefaultTreeAdapterMap['element']): {
+  src: Token.Attribute | undefined
+  sourceCodeLocation: Token.Location | undefined
   isModule: boolean
   isAsync: boolean
 } {
-  let src: AttributeNode | undefined
+  let src: Token.Attribute | undefined
+  let sourceCodeLocation: Token.Location | undefined
   let isModule = false
   let isAsync = false
-  for (let i = 0; i < node.props.length; i++) {
-    const p = node.props[i]
-    if (p.type === NodeTypes.ATTRIBUTE) {
-      if (p.name === 'src') {
-        src = p
-      } else if (p.name === 'type' && p.value && p.value.content === 'module') {
-        isModule = true
-      } else if (p.name === 'async') {
-        isAsync = true
-      }
+  for (const p of node.attrs) {
+    if (p.name === 'src') {
+      src = p
+      sourceCodeLocation = node.sourceCodeLocation?.attrs!['src']
```

Comment:
When there is duplicate attributes, the first one should be selected.
```suggestion
      if (!src) {
        src = p
        sourceCodeLocation = node.sourceCodeLocation?.attrs!['src']
      }
```

Reply:
We only need to transform the first one, no? Subsequent duplicates can be ignored

Reply:
Oh, I see what you mean... nvm. Good catch.